framework/src/service/ServiceHooks.cpp

```diff
@@ -75,10 +74,8 @@ void ServiceHooks::RemovedService(const ServiceReference<ServiceListenerHook>& /
 
 void ServiceHooks::Open()
 {
-  {
-    auto l = this->Lock(); US_UNUSED(l);
-    listenerHookTracker.reset(new ServiceTracker<ServiceListenerHook>(GetBundleContext(), this));
-  }
+  auto l = this->Lock(); US_UNUSED(l);
+  listenerHookTracker.reset(new ServiceTracker<ServiceListenerHook>(GetBundleContext(), this));
```

Comment:
Changing the lock scope in this way is not a good idea because the `->Open()` call below assumes that it is called outside of any lock and will subsequently call into some user call back function.

But I believe that you are right and this code is not properly synchronized with the `Close()` function, because the `listenerHookTracker` unique ptr could have been reset  just before we try to dereference it and call `Open()`. What I think we should do, is to use a `std::shared_ptr<ServiceTracker<ServiceListenerHook>>` instead of a `unique_ptr` for the `listenerHookTracker` member.Then we can keep a local `shared_ptr` inside the `Open()` call and use it outside of the lock scope.